tool/tctl/common/scoped_token_command.go

```diff
@@ -0,0 +1,365 @@
+				"failed to create scoped token (%q already exists), please use another name",
+				tokenName,
+			)
+		}
+		return trace.Wrap(err, "creating scoped token")
+	}
+
+	// Print token information formatted with JSON, YAML, or just print the raw token.
+	switch c.format {
+	case teleport.JSON, teleport.YAML:
+		expires := time.Now().Add(c.ttl)
+		tokenInfo := map[string]any{
+			"token":        tokenName,
+			"roles":        roles,
+			"scope":        tok.GetScope(),
+			"assign_scope": tok.GetSpec().GetAssignedScope(),
+			"expires":      expires,
+		}
+
+		var (
+			data []byte
+			err  error
+		)
+		if c.format == teleport.JSON {
+			data, err = json.MarshalIndent(tokenInfo, "", " ")
+		} else {
+			data, err = yaml.Marshal(tokenInfo)
+		}
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		fmt.Fprint(c.Stdout, string(data))
+
+		return nil
+	case teleport.Text:
+		fmt.Fprintln(c.Stdout, tokenName)
+		return nil
+	}
+
+	// Calculate the CA pins for this cluster. The CA pins are used by the
+	// client to verify the identity of the Auth Server.
+	localCAResponse, err := client.GetClusterCACert(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	caPins, err := tlsca.CalculatePins(localCAResponse.TLSCA)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Get list of auth servers. Used to print friendly signup message.
+	authServers, err := client.GetAuthServers()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(authServers) == 0 {
+		return trace.BadParameter("this cluster has no auth servers")
+	}
+
+	// Print signup message.
+	authServer := authServers[0].GetAddr()
+
+	pingResponse, err := client.Ping(ctx)
+	if err != nil {
+		slog.DebugContext(ctx, "unable to ping auth client", "error", err)
+	}
+
+	if err == nil && pingResponse.GetServerFeatures().Cloud {
+		proxies, err := client.GetProxies()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if len(proxies) != 0 {
+			authServer = proxies[0].GetPublicAddr()
+		}
+	}
+
+	return nodeMessageTemplate.Execute(c.Stdout, map[string]any{
+		"token":        tokenName,
+		"scope":        tok.GetScope(),
+		"assign_scope": tok.GetSpec().GetAssignedScope(),
+		"roles":        strings.ToLower(roles.String()),
+		"minutes":      int(c.ttl.Minutes()),
+		"ca_pins":      caPins,
+		"auth_server":  authServer,
+	})
+}
+
+// Del is called to execute "scoped tokens del ..." command.
+func (c *ScopedTokensCommand) Del(ctx context.Context, client *authclient.Client) error {
+	if c.value == "" {
+		return trace.BadParameter("Need an argument: token")
+	}
+	if err := client.DeleteScopedToken(ctx, c.value); err != nil {
+		return trace.Wrap(err)
+	}
+	fmt.Fprintf(c.Stdout, "Token %s has been deleted\n", c.value)
+	return nil
+}
+
+// List is called to execute "tokens ls" command.
+func (c *ScopedTokensCommand) List(ctx context.Context, client *authclient.Client) error {
+	if mfaResponse, err := mfa.PerformAdminActionMFACeremony(ctx, client.PerformMFACeremony, true /*allowReuse*/); err == nil {
```

Comment:
How come listing tokens requires admin actions but creating and removing them doesn't?

Reply:
It looks like this does happen with some magic for create and delete, this existed here for the tokens API because it was converted to use pagination and may result in hitting multiple APIs. I'm not sure that this is required here. cc @hugoShaka is that correct?